### PR TITLE
1/n Move Accelerator into strategy - move batch_to_device to strategy

### LINKED ISSUE #10648: [Accelerator refactor] Move Accelerator into Strategy
## Proposed refactor

Part 3 of Accelerator and Plugin refactor #10416 

### Motivation

Moving towards stable version

After step 2 Move Precision Plugin into TTP Precision Plugins should be part of Training Type Plugins #7324
Accelerator is not the routing layer for strategy and precision anymore, optimizer related logic, steps, hooks all moved in to strategy. 

Accelerator only have device information - strategy should own accelerator. Reduce the code complexity and improve code maintainability  

### Pitch
move accelerator into ttp/strategy as device_plugin (similar to checkpoint_io) and updating logic in accelerator-connector, training, loops accordingly

[RFC] Should we have a new name for Accelerator?

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @justusschock @awaelchli @akihironitta @kaushikb11 @ananthsub

### PR DESCRIPTION
**Pull request by four4fish**
move batch_to_device to strategy

## What does this PR do?

Part of #10648

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by rohitgr7 — APPROVED** [review]

File: CHANGELOG.md [hunk 1]
```diff
@@ -46,7 +46,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Moved ownership of the `PrecisionPlugin` into `TrainingTypePlugin` and updated all references ([#10570](https://github.com/PyTorchLightning/pytorch-lightning/pull/10570))
 
 
--
+- Moved `batch_to_device` from `Accelerator` to `TrainingTypePlugin` ([#10649](https://github.com/PyTorchLightning/pytorch-lightning/pull/10649))
```
Comment (tchaton) [inline CHANGELOG.md:49]:
```suggestion
- Moved `batch_to_device` method from `Accelerator` to `TrainingTypePlugin` ([#10649](https://github.com/PyTorchLightning/pytorch-lightning/pull/10649))
```

File: pytorch_lightning/plugins/training_type/dp.py [hunk 2]
```diff
@@ -66,6 +66,19 @@ def setup(self) -> None:
...
+
+        The returned batch is of the same type as the input batch, just
+        having all tensors on the correct device.
+        Args:
+            batch: The batch of samples to move to the correct device
+            device: The target device
+            dataloader_idx: The index of the dataloader to which the batch belongs.
+        """
+        device = device or self.root_device
+        return move_data_to_device(batch, device)
```
Comment (tchaton) [inline pytorch_lightning/plugins/training_type/dp.py:80]:
```suggestion
        return move_data_to_device(batch, device=device or self.root_device)
```

File: pytorch_lightning/plugins/training_type/dp.py [hunk 3]
```diff
@@ -66,6 +66,19 @@ def setup(self) -> None:
         self.model_to_device()
         self._model = self._setup_model(LightningParallelModule(self._model))
 
+    def batch_to_device(self, batch: Any, device: Optional[torch.device] = None, dataloader_idx: int = 0) -> Any:
+        """Moves the batch to the correct device.
+
+        The returned batch is of the same type as the input batch, just
```
Comment (tchaton) [inline pytorch_lightning/plugins/training_type/dp.py:72]: This could be re-formulated. Doesn't sound good.

File: pytorch_lightning/plugins/training_type/dp.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/plugins/training_type/dp.py:80]: Why we don't apply `_apply_batch_transfer_handler` with dp? I believe the `on_before_batch_transfer` could still be applied.

```
        model = self.lightning_module
        device = device or self.root_device
        if model is not None:
            return model._apply_batch_transfer_handler(batch, device=device, dataloader_idx=dataloader_idx)
```

File: pytorch_lightning/plugins/training_type/dp.py [hunk 4]
```diff
@@ -66,6 +66,19 @@ def setup(self) -> None:
...
+        having all tensors on the correct device.
+        Args:
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/dp.py:74]: Space here?

```suggestion

        Args:
```

File: pytorch_lightning/plugins/training_type/training_type_plugin.py [hunk 5]
```diff
@@ -90,6 +91,22 @@ def _setup_optimizer(self, optimizer: Optimizer) -> Optimizer:
         # TODO (@awaelchli): standardize this across all plugins in Lightning and Lite. Related refactor: #7324
         return optimizer
 
+    def batch_to_device(self, batch: Any, device: Optional[torch.device] = None, dataloader_idx: int = 0) -> Any:
+        """Moves the batch to the correct device.
+
+        The returned batch is of the same type as the input batch, just
+        having all tensors on the correct device.
+        Args:
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/training_type_plugin.py:99]:
```suggestion

        Args:
```

**Review by carmocca — APPROVED** [review]

**Comment by awaelchli** [comment]
Note, this does not close #10648. It is only the first part. I will unlink the issue.

**Review by awaelchli — APPROVED** [review]
I agree with this temporary move if it simplifies the refactors. In its final form however, it should be the Accelerator that moves the data to the device.

**Comment by four4fish** [comment]
> I agree with this temporary move if it simplifies the refactors. In its final form however, it should be the Accelerator that moves the data to the device.

@awaelchli  Yeah totally agree. The batch_to_device behavior is different between Strategy not device, so easier to move to strategy for step 3 for now. 
After the structure is done, we could revisite the functions and inheritance of strategies - reduce unnecessary inheritant, move functions to utils and etc.

File: pytorch_lightning/plugins/training_type/dp.py [hunk 6]
```diff
@@ -66,6 +66,18 @@ def setup(self) -> None:
         self.model_to_device()
         self._model = self._setup_model(LightningParallelModule(self._model))
 
+    def batch_to_device(self, batch: Any, device: Optional[torch.device] = None, dataloader_idx: int = 0) -> Any:
```
Comment (daniellepintz) [inline pytorch_lightning/plugins/training_type/dp.py:69]: @four4fish why did you create a `batch_to_device` method in dp.py?

Comment (four4fish) [reply pytorch_lightning/plugins/training_type/dp.py:69]: This reproduced accelerator.py L160-L164. Did you see any issue related to this change @daniellepintz?

Comment (daniellepintz) [reply pytorch_lightning/plugins/training_type/dp.py:69]: Right, but why both in dp.py and strategy.py? No, no issues

Comment (daniellepintz) [reply pytorch_lightning/plugins/training_type/dp.py:69]: Oh I see! thanks. No, no issues

Comment (rohitgr7) [reply pytorch_lightning/plugins/training_type/dp.py:69]: this is a bug, no? we don't need to move batch to any device in case of DP since it's handled by DataParallel already. Right now what's happening is:

```
batch - > batch_gpu_0 -> DataParallel -> batch_1_gpu_0 + batch_1_gpu_1
```
how it should be:
```
batch -> DataParallel -> batch_1_gpu_0 + batch_1_gpu_1
```
so we don't need to transfer it by ourselves.

Comment (ananthsub) [reply pytorch_lightning/plugins/training_type/dp.py:69]: @rohitgr7 what's the issue here? https://github.com/PyTorchLightning/pytorch-lightning/blob/8c07d8bf905e395bbd2142b5df7b185b8e936c41/pytorch_lightning/strategies/dp.py#L73-L83

Comment (rohitgr7) [reply pytorch_lightning/plugins/training_type/dp.py:69]: for DP, do we need to move the batch manually to the device[0]?

Comment (four4fish) [reply pytorch_lightning/plugins/training_type/dp.py:69]: @rohitgr7 @ananthsub  I think technically DP doesn't need batch to be moved to device[0], only gradients. I kept the logic because the old accelerator.py L160-L164 has very intentionally defined `move_batch_to_device` for DP. 
I don't know why, so guessed it probably related to a bug caused by multithreading? 
I think we can test this out and remove it if it's not needed.

Comment (rohitgr7) [reply pytorch_lightning/plugins/training_type/dp.py:69]: Well, technically this is not required. Also tested that if we don't move it to device explictly it is working. But I saw PT example too for confirmation and there they are moving batch to device[0]. If there is a reason why this is required, then I'd love to know why else it could be updated.

Comment (rohitgr7) [reply pytorch_lightning/plugins/training_type/dp.py:69]: I did a quick fix some time ago https://github.com/PyTorchLightning/pytorch-lightning/pull/6098

But looks like missed a test.
